x/twap/client/cli/query.go

```diff
@@ -83,6 +85,149 @@ Example:
+
+Example:
+{{.CommandPrefix}} extra-arithmetic 1 uosmo 1667088000 24h
+{{.CommandPrefix}} extra-arithmetic 1 uosmo 1667088000 1667174400
+`, types.ModuleName),
+		Args: cobra.ExactArgs(4),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// boilerplate parse fields
+			twapArgs, err := twapQueryParseArgs(args)
+			if err != nil {
+				return err
+			}
+			clientCtx, err := client.GetClientQueryContext(cmd)
+			if err != nil {
+				return err
+			}
+
+			quoteDenom, err := getQuoteDenomFromLiquidity(cmd.Context(), clientCtx, twapArgs.PoolId, twapArgs.BaseDenom)
+			if err != nil {
+				return err
+			}
+
+			node, err := clientCtx.GetNode()
+			if err != nil {
+				return err
+			}
+			nodeStatus, err := node.Status(context.Background())
+			if err != nil {
+				return err
+			}
+
+			latestHeight := nodeStatus.SyncInfo.LatestBlockHeight
+			diffAccum := cosmossdk_io_math.LegacyZeroDec()
+			diffTime := types.CanonicalTimeMs(twapArgs.EndTime) - types.CanonicalTimeMs(twapArgs.StartTime)
+
+			// Get twap of 2 days chunk then calculate final twap
+			for twapArgs.StartTime.Before(twapArgs.EndTime) {
+				tempStartTime := twapArgs.StartTime
+				tempEndTime := twapArgs.EndTime
+
+				if twapArgs.EndTime.Add(-time.Hour * 48).After(twapArgs.StartTime) {
+					tempStartTime = twapArgs.EndTime.Add(-time.Hour * 48)
+				}
+
+				height, err := findBlockByTime(clientCtx, tempEndTime.UTC(), latestHeight)
+				if err != nil {
+					return err
+				}
+
+				clientCtx, err = getClientQueryContextFromHeight(cmd, height)
+				if err != nil {
+					return err
+				}
+
+				queryClient := queryproto.NewQueryClient(clientCtx)
+
+				res, err := queryClient.ArithmeticTwap(cmd.Context(), &queryproto.ArithmeticTwapRequest{
+					PoolId:     twapArgs.PoolId,
+					BaseAsset:  twapArgs.BaseDenom,
+					QuoteAsset: quoteDenom,
+					StartTime:  tempStartTime,
+					EndTime:    &tempEndTime,
+				})
+				if err != nil {
+					return err
+				}
+
+				timeDelta := types.CanonicalTimeMs(tempEndTime) - types.CanonicalTimeMs(tempStartTime)
+				diffAccum = diffAccum.Add(res.ArithmeticTwap.MulInt64(timeDelta))
+				twapArgs.EndTime = twapArgs.EndTime.Add(-time.Hour * 48)
+				latestHeight = height
+			}
+
+			twap := diffAccum.QuoInt64(diffTime)
+
+			return clientCtx.PrintProto(&queryproto.ArithmeticTwapResponse{ArithmeticTwap: twap})
+		},
+	}
+
+	flags.AddQueryFlagsToCmd(cmd)
+
+	return cmd
+}
+
+// Get Client Query Context at a given height
+func getClientQueryContextFromHeight(cmd *cobra.Command, height int64) (client.Context, error) {
+	ctx := client.GetClientContextFromCmd(cmd)
+	ctx = ctx.WithHeight(height)
+	return client.ReadPersistentCommandFlags(ctx, cmd.Flags())
+}
+
+// Try to find the block containing the given timestamp
+func findBlockByTime(clientCtx client.Context, time time.Time, currentHeight int64) (int64, error) {
+	// Currently we hardcode this value
+	// TODO: Fetch block time from somewhere
+	blockTime := 5.84
```

Comment:
Hrmmm, I'm not sure if the result that came from this hard coded value would have any significant meaning 🤔

Reply:
I take this value from mintscan, its a way for me to find past blocks based on the current time

Reply:
Can we programmatically fetch it from somewhere? E.g. take 10 samples in the past 1000 heights and average them out